File: examples2d/soft_plasticity2.rs
//! Plasticity (2D): elastic cells whose rest shape flows once their strain exceeds a yield
//! (`SoftBodyMaterial::plastic_yield` / `plastic_creep`), shown on squares hit by disks, a
//! stamped clay slab, a creeping column and clay disks splatting on a wall.

use rapier_testbed2d::TestbedViewer;
use rapier2d::prelude::*;

pub async fn run(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    let mut world = PhysicsWorld::new();

    /*
     * Ground, a wall for the clay disks, and a shelf for the press.
     */
    world.insert(
        RigidBodyBuilder::fixed().translation(Vector::new(0.0, -0.5)),
        ColliderBuilder::cuboid(16.0, 0.5),
    );
    world.insert(
        RigidBodyBuilder::fixed().translation(Vector::new(13.0, 2.0)),
        ColliderBuilder::cuboid(0.2, 2.0).friction(0.8),
    );

    let clay = |young: Real, plastic_yield: Real, plastic_creep: Real| SoftBodyMaterial {
        young_modulus: young,
        poisson_ratio: 0.35,
        elastic_damping_ratio: 1.0,
        plastic_yield,
        plastic_creep,
        deformation_damping: 4.0,
        ..Default::default()
    };

    /*
     * The yield ladder: identical squares, from purely elastic (left) to very plastic (right),
     * each hit by the same heavy disk.
     */
    for (i, plastic_yield) in [0.0, 0.2, 0.08, 0.02].into_iter().enumerate() {
        let x = -13.0 + i as Real * 2.4;
        let square = SoftBodyBuilder::grid(Vector::new(x, 0.75), Vector::splat(0.75), 6, 6)
            .cell_model(SoftBodyCellModel::Corotational)
            .material(clay(1.0e4, plastic_yield, 20.0))
            .particle_mass(0.1)
            .surface_collider(ColliderBuilder::ball(0.1).friction(0.8));
        world.insert_soft_body(square);
        world.insert(
            RigidBodyBuilder::dynamic().translation(Vector::new(x, 6.0)),
            ColliderBuilder::ball(0.4).density(5.0),
        );
    }

    /*
     * A clay slab stamped by a kinematic press: the imprints stay after the press lifts.
     */
    let slab = SoftBodyBuilder::grid(Vector::new(0.0, 0.5), Vector::new(3.0, 0.5), 25, 5)
        .cell_model(SoftBodyCellModel::Corotational)
        .material(clay(3.0e4, 0.02, 50.0))
        .particle_mass(0.1)
        .surface_collider(ColliderBuilder::ball(0.12).friction(0.8));
    world.insert_soft_body(slab);
    let press_rest = Vector::new(-2.0, 2.4);
    let (press, _) = world.insert(
        RigidBodyBuilder::kinematic_position_based().translation(press_rest),
        ColliderBuilder::cuboid(0.4, 0.4)
            .rotation(core::f32::consts::FRAC_PI_4)
            .friction(0.5),
    );

    /*
     * Two soft columns under their own weight: the elastic one stands, the plastic one creeps,
     * leans and collapses into a heap (a body whose cells keep flowing is kept awake).
     */
    for (i, plastic_yield) in [0.0, 0.04].into_iter().enumerate() {
        let x = 5.5 + i as Real * 1.5;
        let column = SoftBodyBuilder::grid(Vector::new(x, 1.2), Vector::new(0.3, 1.2), 3, 12)
            .cell_model(SoftBodyCellModel::Corotational)
            .material(SoftBodyMaterial {
                young_modulus: 2.0e3,
                poisson_ratio: 0.35,
                elastic_damping_ratio: 1.0,
                plastic_yield,
                plastic_creep: 0.5,
                deformation_damping: 4.0,
                ..Default::default()
            })
            .particle_mass(0.2)
            .surface_collider(ColliderBuilder::ball(0.1).friction(1.0));
        world.insert_soft_body(column);
    }

    /*
     * Clay disks thrown at the wall.
     */
    let n = 24;
    let indices: Vec<[u32; 2]> = (0..n).map(|i| [i as u32, ((i + 1) % n) as u32]).collect();
    for i in 0..3 {
        let center = Vector::new(11.5 - i as Real * 1.5, 1.0 + i as Real * 0.5);
        let vertices: Vec<Vector> = (0..n)
            .map(|k| {
                let a = k as Real / n as Real * core::f32::consts::TAU;
                center + Vector::new(a.cos(), a.sin()) * 0.5
            })
            .collect();
        if let Some(disk) = SoftBodyBuilder::volumetric(&vertices, &indices, 0.15) {
            let disk = disk
                .cell_model(SoftBodyCellModel::Corotational)
                .material(clay(1.0e4, 0.03, 60.0))
                .particle_mass(0.05)
                .surface_collider(ColliderBuilder::ball(0.075).friction(0.8));
            let handle = world.insert_soft_body(disk);
            let sb = &mut world.soft_bodies[handle];
            for k in 0..sb.num_particles() {
                sb.set_particle_velocity(k, Vector::new(10.0, 1.0));
            }
        }
    }

    viewer.set_world(&mut world);
    viewer.look_at(Vec2::new(0.0, 3.0), 30.0);

    let mut t: Real = 0.0;
    while viewer.render_frame(&mut world).await {
        if viewer.simulating() {
            t += world.integration_parameters.dt;
            // The press stamps a new spot every 3 seconds: down for a second, up for a second,
            // then slides to the next spot.
            let period = 3.0;
            let cycle = (t / period).floor();
            let phase = t - cycle * period;
            let x = press_rest.x + (cycle % 5.0) * 1.0;
            let depth = 1.0;
            let y = press_rest.y
                - if phase < 1.0 {
                    depth * phase
                } else if phase < 2.0 {
                    depth * (2.0 - phase)
                } else {
                    0.0
                };
            world.bodies[press].set_next_kinematic_translation(Vector::new(x, y));
            world.step();
        }
    }
    Ok(())
}

File: c/testbed/examples2d/soft_plasticity2.c
/* Port of examples2d/soft_plasticity2.rs. */
#include "testbed.h"
#include "rapier_helpers.h"
#include "rapier_math.h"

static R2SoftBodyMaterial clay(R2Real young, R2Real plasticYield, R2Real plasticCreep) {
    R2SoftBodyMaterial material = r2DefaultSoftBodyMaterial();
    material.youngModulus = young;
    material.poissonRatio = .35;
    material.elasticDampingRatio = 1;
    material.plasticYield = plasticYield;
    material.plasticCreep = plasticCreep;
    material.deformationDamping = 4;
    return material;
}

void tbSoftPlasticity2(Testbed *testbed) {
    R2World *world = r2NewWorld();
    {
        R2RigidBodyDesc rigidBody = r2FixedRigidBodyDesc();
        rigidBody.position.translation = r2Vector(0, -0.5);
        rigidBody.canSleep = !testbed->noSleep;
        R2ColliderDesc collider = r2CuboidColliderDesc(r2Vector(16, 0.5));
        R2RigidBodyHandle rigidBodyHandle = r2InsertRigidBody(world, &rigidBody);
        r2InsertCollider(rigidBodyHandle, &collider);
    }
    {
        R2RigidBodyDesc rigidBody = r2FixedRigidBodyDesc();
        rigidBody.position.translation = r2Vector(13, 2);
        rigidBody.canSleep = !testbed->noSleep;
        R2ColliderDesc collider = r2CuboidColliderDesc(r2Vector(0.2, 2));
        collider.friction = .8;
        R2RigidBodyHandle rigidBodyHandle = r2InsertRigidBody(world, &rigidBody);
        r2InsertCollider(rigidBodyHandle, &collider);
    }
    /* Yield ladder: elastic to increasingly plastic, hit by identical balls. */
    const R2Real yields[] = {0, .2, .08, .02};
    for (size_t i = 0; i < TB_COUNT(yields); ++i) {
        const R2Real x = -13 + i * 2.4;
        R2SoftBodyDesc square = r2GridSoftBodyDesc(r2Vector(x, 0.75), r2Vector(0.75, 0.75), 6, 6);
        square.cellModel = R2_SOFT_CELL_COROTATIONAL;
        square.particleMass = .1;
        square.canSleep = !testbed->noSleep;
        {
            R2SoftBodyMaterial material = clay(1.0e4, yields[i], 20);
            square.material = material;
        }
        {
            R2ColliderDesc surface = r2BallColliderDesc(.1);
            surface.friction = .8;
            square.collider = surface;
        }
        r2InsertSoftBody(world, &square);

        {
            R2RigidBodyDesc rigidBody = r2DynamicRigidBodyDesc();
            rigidBody.position.translation = r2Vector(x, 6);
            rigidBody.canSleep = !testbed->noSleep;
            R2ColliderDesc collider = r2BallColliderDesc(.4);
            collider.density = 5;
            R2RigidBodyHandle rigidBodyHandle = r2InsertRigidBody(world, &rigidBody);
            r2InsertCollider(rigidBodyHandle, &collider);
        }
    }
    /* Clay slab stamped by a kinematic press. */
    R2SoftBodyDesc slab = r2GridSoftBodyDesc(r2Vector(0, 0.5), r2Vector(3, 0.5), 25, 5);
    slab.cellModel = R2_SOFT_CELL_COROTATIONAL;
    slab.particleMass = .1;
    slab.canSleep = !testbed->noSleep;
    {
        R2SoftBodyMaterial material = clay(3.0e4, .02, 50);
        slab.material = material;
    }
    {
        R2ColliderDesc surface = r2BallColliderDesc(.12);
        surface.friction = .8;
        slab.collider = surface;
    }
    r2InsertSoftBody(world, &slab);

    const R2Vector pressRest = r2Vector(-2, 2.4);
    R2RigidBodyHandle press;
    {
        R2RigidBodyDesc rigidBody = r2KinematicPositionBasedRigidBodyDesc();
        rigidBody.position.translation = pressRest;
        rigidBody.canSleep = !testbed->noSleep;
        R2ColliderDesc collider = r2CuboidColliderDesc(r2Vector(0.4, 0.4));
        collider.position.rotation = r2Rotation(R2_PI / 4);
        collider.friction = .5;
        press = r2InsertRigidBody(world, &rigidBody);
        r2InsertCollider(press, &collider);
    }
    /* Elastic and creeping columns under their own weight. */
    for (int i = 0; i < 2; ++i) {
        const R2Real x = 5.5 + i * 1.5;
        R2SoftBodyDesc column = r2GridSoftBodyDesc(r2Vector(x, 1.2), r2Vector(0.3, 1.2), 3, 12);
        column.cellModel = R2_SOFT_CELL_COROTATIONAL;
        column.particleMass = .2;
        column.canSleep = !testbed->noSleep;
        {
            R2SoftBodyMaterial material = clay(2.0e3, i == 0 ? 0 : .04, .5);
            column.material = material;
        }
        {
            R2ColliderDesc surface = r2BallColliderDesc(.1);
            surface.friction = 1;
            column.collider = surface;
        }
        r2InsertSoftBody(world, &column);
    }
    /* Volumetric clay disks thrown at the wall. */
    const size_t n = 24;
    uint32_t indices[24][2];
    for (uint32_t i = 0; i < n; ++i) {
        indices[i][0] = i;
        indices[i][1] = (i + 1) % n;
    }
    for (int i = 0; i < 3; ++i) {
        const R2Vector center = r2Vector(11.5 - i * 1.5, 1 + i * .5);
        R2Vector vertices[24];
        for (size_t k = 0; k < n; ++k) {
            const R2Real a = (R2Real)k / n * 2 * R2_PI;
            vertices[k] = r2VectorAdd(center, r2Vector(.5 * cos(a), .5 * sin(a)));
        }

        R2VolumeMeshParameters ballMeshing = r2NewVolumeMeshParameters(.15);
        R2SoftBodyDesc ball = r2VolumetricSoftBodyDesc(
            (R2VectorView){vertices, n}, (R2SurfaceElementView){(const R2Edge *)&indices[0][0], n},
            ballMeshing);
        {
            ball.cellModel = R2_SOFT_CELL_COROTATIONAL;
            R2SoftBodyMaterial material = clay(1.0e4, .03, 60);
            ball.material = material;
            ball.particleMass = .05;
            ball.canSleep = !testbed->noSleep;
            {
                R2ColliderDesc surface = r2BallColliderDesc(.075);
                surface.friction = .8;
                ball.collider = surface;
            }
            R2SoftBodyHandle handle = r2InsertSoftBody(world, &ball);

            size_t count = r2SoftBody_NumParticles(handle);
            for (size_t k = 0; k < count; ++k) {
                r2SoftBody_SetParticleVelocity(handle, k, r2Vector(10, 1));
            }
        }
    }

    tbCamera2(testbed, 0, 3, 30);

    tbSetWorld(testbed, world);
    R2Real t = 0;
    testbed->snapshotSupported = 0;

    while (tbRenderFrame(testbed, &world)) {
        if (tbSimulating(testbed)) {
            R2Real dt = r2TimeStep(world);
            t += dt;
            /* One second down, one second up, then move to the next stamp. */
            const R2Real period = 3;
            const R2Real cycle = floor(t / period);
            const R2Real phase = t - cycle * period;
            const R2Real x = pressRest.x + fmod(cycle, 5);
            const R2Real depth = 1;
            const R2Real y = pressRest.y - (phase < 1   ? depth * phase
                                            : phase < 2 ? depth * (2 - phase)
                                                        : 0);

            r2RigidBody_SetNextKinematicTranslation(press, r2Vector(x, y));
            r2Step(world, NULL, NULL);
        }
    }
    r2FreeWorld(world);
}
